torch/utils/data/dataset.py

```diff
@@ -260,6 +265,125 @@ def __len__(self):
         return len(self.indices)
 
 
+class ChunkDataset(IterableDataset):
+    r"""Dataset which uses hierarchical sampling for efficient data reading.
+
+
+    ``ChunkDataset`` is a stateful dataset that supports hierarchical sampling and
+    efficient reading through chunks. A chunk is selected based on a sampling strategy
+    (first level) and its content is shuffled (second level) before producing
+    a batch or an example.
+
+    In this context, chunks could be files, contents of a folder,
+    sections of large text file, etc. In distributed training, to eliminate different workers reading
+    uneven amount of data, the chunks need to be roughly the same size.
+    Please consult ChunkDataReader for more details.
+
+    ``ChunkDataset`` extends ``IterableDataset`` because the length of the dataset
+    (or simply the size) is unknown. Only the number of chunks is known for the dataset.
+
+    In a distributed setup, each `DataLoader` worker has an instance of ``ChunkDataset`` that uses
+    ``DistributedChunkSampler`` sampler to select chunks based on their global worker rank.
+    Once a chunk index has been selected, it is passed to ``ChunkDataReader`` that will load
+    data into the ``ChunkDataset`` internal cache. Only then ``ChunkDataset`` returns batches.
+
+    Arguments:
+        chunk_sampler (DistributedSampler or DistributedChunkSampler): Draw indices for chunks.
+            Typically used to split data amongst dataloader workers (first level of sampling)
+        chunk_reader (ChunkDataReader): Specialized reader for a given input type
+        shuffle_cache (bool): Setting `True` trigger shuffling of the internal chunk cache
+            (second level of sampling) (default: `True`)
+    """
+
+    def __init__(self, chunk_sampler, chunk_reader, shuffle_cache=True):
+        super(ChunkDataset, self).__init__()
+        assert isinstance(chunk_sampler, DistributedSampler) or isinstance(
+            chunk_sampler, DistributedChunkSampler), 'sampler must be a `DistributedSampler` or `DistributedChunkSampler`'
+        assert callable(chunk_reader), 'chunk_reader must be `callable()` and return a container with data'
+        assert isinstance(shuffle_cache, bool), 'shuffle_cache must be a `bool`'
+
+        self.chunk_sampler = chunk_sampler
+        self.chunk_reader = chunk_reader
+        self.shuffle_cache = shuffle_cache
+
+        # Internal state
+        self._chunk_sampler_iter = iter(self.chunk_sampler)
+        self._cache = []
+        self._batch_size = 1
+        self._min_cache = 1000
+
+    def __iter__(self):
+        return self
+
+    def __len__(self):
+        raise NotImplementedError
```

Comment:
Nit: please add some simple comment on why this returns NotImplementedError.

Reply:
At the class documentation there is a paragraph explaining the reason, but I will add a brief comment here too

> ``ChunkDataset`` extends ``IterableDataset`` because the length of the dataset
    (or simply the size) is unknown. Only the number of chunks is known for the dataset.

In fact, we could delete the `__len__` method altogether as it is not part of neither `IterableDataset` and `ChunkDataset`. I have added this method as an additional protection